crates/turbo-tasks-memory/src/task_statistics.rs

```diff
@@ -0,0 +1,82 @@
+use std::sync::{Arc, OnceLock};
+
+use dashmap::DashMap;
+use serde::{ser::SerializeMap, Serialize, Serializer};
+use turbo_tasks::{registry, FunctionId};
+
+/// An API for optionally enabling, updating, and reading aggregated statistics.
+#[derive(Default)]
+pub struct TaskStatisticsApi {
+    inner: OnceLock<Arc<TaskStatistics>>,
+}
+
+impl TaskStatisticsApi {
+    pub fn enable(&self) -> &Arc<TaskStatistics> {
+        // ignore error: potentially already initialized, that's okay
+        let _ = self.inner.set(Arc::new(TaskStatistics {
+            inner: DashMap::new(),
+        }));
+        self.inner.get().unwrap()
+    }
+
+    pub fn is_enabled(&self) -> bool {
+        self.inner.get().is_some()
+    }
+
+    // Calls `func` if statistics have been enabled (via
+    // [`TaskStatisticsApi::enable`]).
+    pub fn map<T>(&self, func: impl FnOnce(&Arc<TaskStatistics>) -> T) -> Option<T> {
+        self.get().map(func)
+    }
+
+    // Calls `func` if statistics have been enabled (via
+    // [`TaskStatisticsApi::enable`]).
+    pub fn get(&self) -> Option<&Arc<TaskStatistics>> {
+        self.inner.get()
+    }
+}
+
+/// A type representing the enabled state of [`TaskStatisticsApi`]. Implements
+/// [`serde::Serialize`].
+pub struct TaskStatistics {
+    inner: DashMap<FunctionId, TaskFunctionStatistics>,
```

Comment:
```suggestion
    inner: DashMap<FunctionId, TaskFunctionStatistics, rustc_hash::FxBuildHasher>,
```

`DashMap` uses the default hasher, so it can be slow. I think if this type stores stats of all tasks, it can be an issue.